Expand SPIR-V features

DialecticalMaterialist opened the pull request:
Currently this adds support for every SPIR-V Extension and Capability. You enable them by setting which ones you want in a featureSet then assigning it to the cpu_features_add in your target. There is also now support for setting some Execution Modes via the calling convention. Following is an example setting the SPIR-V version to 1.6, enabling the SPV_EXT_mesh_shader extension and enabling the MeshShadingEXT capability it also adds the required Execution Modes via the calling convention. 

**build.zig**
```
const std = @import("std");
pub fn build(b: *std.Build) void {
    const spirv_target = b.resolveTargetQuery(.{
        .cpu_arch = .spirv64,
        .cpu_model = .{ .explicit = &std.Target.spirv.cpu.generic },
        .cpu_features_add = std.Target.spirv.featureSet(&[_]std.Target.spirv.Feature{ .mesh_shading_ext, .v1_6 }),
        .os_tag = .opengl,
        .ofmt = .spirv,
        .abi = .none,
    });

    const shader = b.addObject(.{
        .name = "shader",
        .root_module = b.createModule(.{
            .root_source_file = b.path("shader.zig"),
            .target = spirv_target,
            .optimize = .ReleaseFast,
        }),
        .use_llvm = false,
    });

    b.getInstallStep().dependOn(&b.addInstallFile(shader.getEmittedBin(), "../shader.spv").step);
}
```
**shader.zig**

```
export fn main_mesh() callconv(.{ .spirv_mesh = .{
    .stage_output = .output_triangles,
    .max_primitives = 10,
    .max_vertices = 30,
} }) void {}
```
**shader.spv**

```
; SPIR-V
; Version: 1.6
; Generator: Khronos; 41
; Bound: 34
; Schema: 0
               OpCapability Shader
               OpCapability Matrix
               OpCapability Int64
               OpCapability MeshShadingEXT
               OpCapability Int8
               OpCapability Int16
               OpExtension "SPV_EXT_mesh_shader"
               OpMemoryModel Logical GLSL450
               OpEntryPoint MeshEXT %25 "main_mesh"
               OpExecutionMode %25 OutputVertices 30
               OpExecutionMode %25 OutputPrimitivesEXT 10
               OpExecutionMode %25 OutputTrianglesEXT
               OpSourceExtension "zig_errors:"
               OpSource Zig 0
               OpName %void "void"
               OpName %shader_main_mesh "shader.main_mesh"
       %void = OpTypeVoid
         %26 = OpTypeFunction %void
%shader_main_mesh = OpFunction %void None %26
          %4 = OpLabel
               OpReturn
          %5 = OpLabel
               OpUnreachable
               OpFunctionEnd
         %25 = OpFunction %void None %26
         %32 = OpLabel
         %33 = OpFunctionCall %void %shader_main_mesh
               OpReturn
               OpFunctionEnd
               
```

As we can see the version is correctly set, the extension is there and the capability as well as the required Execution Modes.

alichraghi commented:
Actually, i think we should remove `OpCapability/OpExtension` instructions support from assembler and instead populate `std.Target.spirv.Feature`.

CC @Snektron

alichraghi commented on src/codegen/spirv/Module.zig:
```diff
@@ -19,6 +19,161 @@ const Id = spec.Id;
 
 const Section = @import("Section.zig");
 
+pub const Extensions = enum {
```
This must be auto-generated by `tools/gen_spirv_spec.zig` and go into `src/codegen/spirv/spec.zig`

DialecticalMaterialist replied:
Problem is that Extensions are not defined in the json file that gen_spirv_spec uses and some of the Extensions are scattered over multiple places, I got some from here: https://github.com/KhronosGroup/SPIRV-Registry and others from here https://github.com/intel/llvm/tree/sycl/sycl/doc/design/spirv-extensions and some from other places I have yet to find a place where all are gathered.

alichraghi commented on tools/update_cpu_features.zig:
```diff
@@ -1055,6 +1055,7 @@ const targets = [_]ArchTarget{
         },
         .branch_quota = 2000,
         .extra_features = &.{
```
Perhaps we can avoid writing these manually as well and extract them from `spec.zig`?

alichraghi commented on src/codegen/spirv/Module.zig:
```diff
@@ -360,22 +514,737 @@ pub fn finalize(self: *Module, a: Allocator) ![]Word {
     try self.addCapability(.int8);
     try self.addCapability(.int16);
 
+    // Core capabilities
```
These could be automated too! Although, i don't see it as a requirement here.

DialecticalMaterialist commented:
Ah wait I dumb. I now see that https://raw.githubusercontent.com/KhronosGroup/SPIRV-Headers/refs/heads/main/include/spirv/unified1/spirv.core.grammar.json contains the capabilities with their extensions. Okay then that can be automated with gen_spirv_spec and update_cpu_features

DialecticalMaterialist commented:
> Yes. Note that i reduced the instructions sets to only core, opencl and glsl so don't bother supporting others.

What do you mean by this?

DialecticalMaterialist commented:
That is on your spv4 branch yes? I'm going to add modify gen_spirv_spec soon to generate Extensions and Capabilities which has version and requirements, should I rebase on your fork?

DialecticalMaterialist commented:
Would  I need to move this pr to your fork then?

DialecticalMaterialist commented:
@alichraghi would you know how we can access spec.zig in update_cpu_features.zig?

alichraghi commented:
Hmm that would be tricky because we now have to pass `--dep` which is not desirable so we probably have no other choice but adding an step named like `update-cpu-features` to `build.zig`.

CC @alexrp 

TLDR; We need to import `src/codegen/spirv/spec.zig` in `update_cpu_features.zig` to auto-generate SPIR-V CPU features

DialecticalMaterialist commented:
How would one do it with --dep? So that I can continue working on update_cpu_features.zig will waiting for proper way.

DialecticalMaterialist commented:
Thanks, good news is that this is only thing left for it work. I compiled latest commit and it works, it emits correct OpCapability and OpExtension, only stuff that don't work are some capabilities that I did not define in spirv.Feature, from the first commits. But now when update_cpu_features works it should be done

DialecticalMaterialist commented:
@alichraghi Would you know how one would fix the "global variable contains reference to comptime var" error when compiling update_cpu_features.zig? Is this way of using compile time not allowed how would one otherwise implement this?

DialecticalMaterialist commented:
@alichraghi do you know why openGL does not define bit sizes of C types in std.Target and instead just panics? Should it not have the same as mesa3d?

Commit:
spirv: remove `OpCapability`/`OpExtension` support from assembler

All known capabilities and extensions are now added into `std.Target`
as features.

DialecticalMaterialist commented:
@alichraghi I have rebased now and removed the instructions from assembler. It compiles for me but it still probably fails CI checks due to update_cpu_features.zig dependency on spirv spec.zig

DialecticalMaterialist commented:
@alichraghi Why did you remove OpExecution support from inline assembly? And beside old implementation was wrong since it didn't allow setting execution mode more than once. This is wrong because some stages require that you set more than one, like Mesh shaders or Geometry shaders, etc. ExecutionMode is not XOR it allows multiple. See here in spec https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#Execution_Mode it states "It is invalid to apply the same execution mode more than once to any entry point unless explicitly allowed below for a specific execution mode.". The exec_mode in EntryPoint struct should really be an array. Also we executionMode function gpu.zig which doesn't work because of change. I think it is best if we read support for inline assembly ExecutionMode since currently there is no alternative to set it besides hardcoding for specific calling conventions like I did for mesh shader but this is not good because this severely limits what you can do. I think while we don't have language support for certain SPIR-V features they should be available in inline assembly

DialecticalMaterialist commented:
I propose that exec_mode gets moved out of entry_points in accordance to section 2.4 of the SPIR-V Specification.

alichraghi commented:
It's meant to be replaced with proper support via calling convention. The problem is those parameters are too large to fit `std.builtin.CallingConvention` so this probably needs to be discussed later. And I don't think it's worth to bring back `OpExecutionMode` to assembler. We already lack language-level support for basic opaque types like `OpTypeImage`.

DialecticalMaterialist commented:
But this makes mesh and task shaders impossible, I think that we should continue support OpExecutionMode while waiting for a new method such as via calling convention

DialecticalMaterialist commented:
It seems like InternPool stores a packed one which is only 18 bits, if it was 64 then it wouldn't be that big problem, but if it is only 18 idk how. I'll keep looking and see if I can find a solution

DialecticalMaterialist commented:
But it seems like the packed calling convention can be increased to 27 bits since it is used in `TypeFunction.Flags` which has 9 unused bits to get it up to 32 bits in a packed struct

alichraghi commented:
@DialecticalMaterialist Lets keep CallingConvention stuff out of this PR. and regarding the `spec.zig` dependency, I don't think there's a better solution than adding a new step to `build.zig`

Commit:
spirv: set execution mode via calling convention

DialecticalMaterialist commented:
I think it is possible to get some support with CC, only problem is that we are currently limited by bits. And how would I add the dep to build.zig? I'm new to this codebase so I don't really know how all the stuff work yet, would appreciate if you or someone else could help with that

alichraghi commented:
I can take over the rest. don't worry.

DialecticalMaterialist commented:
@alichraghi I managed to get CC to work with the recent commit.

**build.zig**
```
const std = @import("std");
pub fn build(b: *std.Build) void {
    const spirv_target = b.resolveTargetQuery(.{
        .cpu_arch = .spirv64,
        .cpu_model = .{ .explicit = &std.Target.spirv.cpu.generic },
        .cpu_features_add = std.Target.spirv.featureSet(&[_]std.Target.spirv.Feature{ .mesh_shading_ext, .v1_6 }),
        .os_tag = .opengl,
        .ofmt = .spirv,
        .abi = .none,
    });

    const shader = b.addObject(.{
        .name = "shader",
        .root_module = b.createModule(.{
            .root_source_file = b.path("shader.zig"),
            .target = spirv_target,
            .optimize = .ReleaseFast,
        }),
        .use_llvm = false,
    });

    b.getInstallStep().dependOn(&b.addInstallFile(shader.getEmittedBin(), "../shader.spv").step);
}
```
**shader.zig**

```
export fn main_mesh() callconv(.{ .spirv_mesh = .{
    .stage_output = .output_triangles,
    .max_primitives = 1,
    .max_vertices = 3,
} }) void {}
```
**shader.spv**
```
; SPIR-V
; Version: 1.6
; Generator: Khronos; 41
; Bound: 34
; Schema: 0
               OpCapability Shader
               OpCapability Matrix
               OpCapability Int64
               OpCapability MeshShadingEXT
               OpCapability Int8
               OpCapability Int16
               OpExtension "SPV_EXT_mesh_shader"
               OpMemoryModel Logical GLSL450
               OpEntryPoint MeshEXT %25 "main_mesh"
               OpExecutionMode %25 OutputVertices 3
               OpExecutionMode %25 OutputPrimitivesEXT 1
               OpExecutionMode %25 OutputTrianglesEXT
               OpSourceExtension "zig_errors:"
               OpSource Zig 0
               OpName %void "void"
               OpName %shader_main_mesh "shader.main_mesh"
       %void = OpTypeVoid
         %26 = OpTypeFunction %void
%shader_main_mesh = OpFunction %void None %26
          %4 = OpLabel
               OpReturn
          %5 = OpLabel
               OpUnreachable
               OpFunctionEnd
         %25 = OpFunction %void None %26
         %32 = OpLabel
         %33 = OpFunctionCall %void %shader_main_mesh
               OpReturn
               OpFunctionEnd

```
That is a valid mesh shader tested with spirv-val

DialecticalMaterialist commented:
Should I continue adding support for other calling conventions? As I stated before it is still really limited due to only having access to 13 bits lol, maybe one could squeeze out 6 more bits by doing some cursed stuff with incoming_stack_alignment

alichraghi commented:
A kernel execution mode info can't be fit in a `PackedCallingConvention` so it should be moved out of `Flags`. I'll send a PR to your branch soon to continue this and also add the `build.zig` step if you don't mind.

alichraghi committed:
build: add `update-cpu-features` step

this tool now has to be run through build system duo to a dependency on `src/codegen/spirv/spec.zig`

alichraghi commented:
Sorry, Github isn't showing your repo on PR panel to me for some reason. Can you pull [my branch](https://github.com/alichraghi/zig/tree/help-dialect) manually?

DialecticalMaterialist commented:
Sure!

alichraghi commented on lib/std/builtin.zig:
```diff
@@ -346,11 +346,27 @@ pub const CallingConvention = union(enum(u8)) {
+    spirv_task,
+    spirv_mesh: SpirvMeshOptions,
+
+    pub const SpirvKernelOptions = struct { x: u32, y: u32, z: u32 };
+
+    pub const SpirvMeshOptions = struct {
+        const StageOutput = enum(u2) {
+            output_points = 0,
+            output_lines = 1,
+            output_triangles = 2,
+        };
+
+        stage_output: StageOutput = .output_triangles,
+        max_primitives: u32 = 1,
+        max_vertices: u32 = 3,
+    };
```
These should go under `RiscvInterruptOptions`

alichraghi commented on src/codegen/spirv/Module.zig:
```diff
@@ -317,26 +310,113 @@ fn entryPoints(module: *Module) !Section {
+            else => {},
         }
+
+        // if (entry_point.exec_mode == null) {
+        //     switch (target.os.tag) {
+        //         .vulkan, .opengl => |tag| {
+        //             switch (entry_point.exec_model) {
+        //                 .fragment => {
```
Forgot to remove this?

alichraghi reviewed (commented):
Just some nitpicks i forgot to take care of. Also Lets wait for @alexrp and/or @mlugg reviews before pushing to not stress CI.

alichraghi commented:
> I'm going to add some for fragment only u4 so fits in extra. Should I open this PR again after that?

If you mean `Origin[Upper/Lower]Left`, the answer is No because there's only one valid value for each opengl and vulkan targets.

DialecticalMaterialist commented:
No it's the depth thing and if pixel centered integer, tho maybe I should add check that you can't set pixel centered integer on Vulkan

alexrp commented on lib/std/Target.zig:
```diff
@@ -3033,6 +3035,13 @@ pub fn cTypeBitSize(target: *const Target, c_type: CType) u16 {
             .longdouble => return 128,
         },
 
+        .opengl => switch (c_type) {
+            .char => return 8,
+            .short, .ushort => return 16,
+            .int, .uint, .float => return 32,
+            .long, .ulong, .longlong, .ulonglong, .double, .longdouble => return 64,
+        },
```
Where do these values come from?

alexrp replied:
Note that this question still stands even with the changed values. It's important that the source of ABI-defining values like this is known.

alexrp replied:
So the thing is that this is about C ABI. I'm not aware of a C-to-GLSL compiler or something like that. So strictly speaking, these values aren't correct.

But let's take a step back:

> So I think these new values are correct since otherwise you can't compile

Can you elaborate on this? What made filling them out necessary? I wouldn't have expected that to be needed unless you actually use the C types (`c_int`, etc).

alexrp replied:
Could you get me a stack trace of this panic? Or a repro I can run against master I guess. Since we're talking about adding non-standard ABI values, I'd like to properly understand the problem before I sign off on that.

DialecticalMaterialist replied:
Here is an example using the current commit of this branch with those types removed and opengl instead moved to ps3 and cause panic

**build.zig**
```
const std = @import("std");
pub fn build(b: *std.Build) void {
    const spirv_target = b.resolveTargetQuery(.{
        .cpu_arch = .spirv64,
        .cpu_model = .{ .explicit = &std.Target.spirv.cpu.generic },
        .os_tag = .opengl,
        .ofmt = .spirv,
        .abi = .none,
    });

    const shader = b.addObject(.{
        .name = "shader",
        .root_module = b.createModule(.{
            .root_source_file = b.path("shader.zig"),
            .target = spirv_target,
            .optimize = .ReleaseFast,
        }),
        .use_llvm = false,
    });

    b.getInstallStep().dependOn(&b.addInstallFile(shader.getEmittedBin(), "../shader.spv").step);
}
```

**shader.zig**
```
export fn main() callconv(.spirv_vertex) void {
    const a: f32 = 1.0;
    var b: f32 = undefined;
    b = a;
}
```

```
error: thread 57516 panic: specify the C integer and float type sizes for this OS
Unable to dump stack trace: debug info stripped
```

DialecticalMaterialist replied:
This was the reason that I defined them since floating point types depend on that c type function for some reason, I'm pretty sure this is for all platforms. Maybe that should be looked into?

DialecticalMaterialist replied:
Shouldn't it define the types then since OpenCL and Vulkan do?

alichraghi commented on lib/std/Target.zig:
```diff
@@ -1144,7 +1144,7 @@ pub const Cpu = struct {
             ints: [usize_count]usize,
 
-            pub const needed_bit_count = 288;
+            pub const needed_bit_count = 421;
```
My memory might be flushed but was it ever been this way before?
AFAIK the impression was that capabilities/extensions through feature set is the right way we just didn't take time to add all of them w(e are still not %100 of them because they are out of core/glsl/opencl spec).

alexrp replied:
#22889

alichraghi replied:
Ough okay yeah, FWIW i think my past self was wrong because SPIR-V capabilities map very well with the Zig feature set and enabling them through assembly just makes the whole compilation more implicit because we make multiple decisions based on them.
Also currently some caps/extensions can be enabled via both assembly and feature-set so yet another reason to unify them all.

DialecticalMaterialist commented:
Should I resolve the bit size thing?

alichraghi reviewed (approved):
LGTM! Nice work! I think Sema/InternPool changes need to be reviewed by @mlugg tho.

alichraghi commented:
Thanks. Would be great if you reset based on my branch instead of merging though, because the commit history is a mess right now.
```
git reset --hard alichraghi/dm
```

Commit:
Merge branch 'master' into expand-gpu

Commit:
Merge branch 'master' of https://github.com/ziglang/zig into expand-gpu

alexrp commented:
https://codeberg.org/ziglang/zig/pulls/30124
